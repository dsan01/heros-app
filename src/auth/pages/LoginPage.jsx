import { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context'

export const LoginPage = () => {

  const { login } = useContext(AuthContext)
  const navigate = useNavigate()

  const onLogin = () => {
    
    login('David Sanchez')
    
    navigate("/marvel", {
      replace: true
    });
}

  return (
    <div className='container mt-5 mx-auto'>
      <h1 className='text-2xl'> Login</h1>
      <hr className='my-4' />

      <button
        className='w-max bg-blue-500 px-4 py-2 rounded-lg text-white cursor-pointer hover:bg-blue-300'
        onClick={ onLogin }
      > 
        Login
      </button>

    </div>
  )
}
