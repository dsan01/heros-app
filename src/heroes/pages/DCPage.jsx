import React from 'react'
import { HeroList } from '../components'

export const DCPage = () => {
  return (
    <>
      <h2 className='text-3xl font-bangers my-4 text-blue-600'>DC Comics</h2>
      <hr />
      <HeroList publisher={'DC Comics'} />
    </>
  )
}
