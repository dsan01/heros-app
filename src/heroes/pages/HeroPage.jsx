import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faBook, faCalendarDays, faUser } from "@fortawesome/free-solid-svg-icons"

export const HeroPage = () => {

    const { id } = useParams()
    const navigate = useNavigate()

    const hero = getHeroById(id)

    const onNavigateBack = () => {
        navigate(-1);
    }

    if (!hero) {
        return <Navigate to={'/'} />
    }

    return (
        <>
            <div className="grid grid-cols-12 my-5">
                <div className="col-span-4">
                    <img src={`/assets/heroes/${id}.jpg`} alt={hero.superhero} className="rounded-xl shadow-md" />
                </div>
                <div className="col-span-8 px-8 py-4">
                    <h2 className="text-3xl font-semibold"> {hero.superhero} </h2>
                    <div className="mt-4 border p-4 rounded">
                        <p className="mb-1"> <FontAwesomeIcon icon={faBook} className="text-gray-500" /> <span className="font-medium ml-2"> Publicado por: </span> {hero.publisher} </p>
                        <p className="mb-1"> <FontAwesomeIcon icon={faUser}  className="text-gray-500" /> <span className="font-medium ml-2"> Nombre: </span> {hero.alter_ego} </p>
                        <p className="mb-1"> <FontAwesomeIcon icon={faCalendarDays}  className="text-gray-500" /> <span className="font-medium ml-2"> Primera apariciÃ³n: </span> {hero.first_appearance} </p>
                    </div>
                    <button
                        className="w-max bg-blue-500 px-4 py-2 my-4 rounded-lg text-white cursor-pointer hover:bg-blue-300"
                        onClick={onNavigateBack}
                    >
                        Atras
                    </button>
                </div>
            </div>
        </>
    )
}
