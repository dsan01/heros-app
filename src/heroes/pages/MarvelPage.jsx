import { HeroList } from "../components"

export const MarvelPage = () => {
    return (
        <>
            <h2 className='text-3xl font-bangers my-4 text-red-600'>Marvel Comics</h2>
            <hr />
            <HeroList publisher={'Marvel Comics'} />
        </>
    )
}
