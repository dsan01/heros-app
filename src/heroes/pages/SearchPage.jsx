import { useSearchParams } from "react-router-dom";
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../components"
import { getHeroesByName } from "../helpers";


export const SearchPage = () => {

    const [searchParams, setSearchParams] = useSearchParams();
    
    const q = searchParams.get('q') ?? '';
    
    const heroes = getHeroesByName(q)
    
    const { searchtext, onInputChange } = useForm({
        searchtext: q
    });

    const onSearchSubmit = (event) => {
        event.preventDefault()
        console.log('holi')
        setSearchParams({q: searchtext});
    }

    return (
        <>
            <div className="grid grid-cols-12 my-5">
                <h1 className="text-3xl font-semibold col-span-12"> Buscar un heroe </h1>
                <div className="col-span-5 px-4">
                    <h4 className="text-xl py-4 font-medium">Busqueda</h4>
                    <form onSubmit={onSearchSubmit} aria-label="form">
                        <input
                            type="text"
                            className=" shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
                            placeholder="Busca un superheroe"
                            name="searchtext"
                            autoComplete="off"
                            value={searchtext}
                            onChange={onInputChange}
                        />
                        <button
                            className="w-max bg-blue-500 px-4 py-2 my-4 rounded-lg text-white cursor-pointer hover:bg-blue-300"
                        >
                            Buscar
                        </button>
                    </form>
                </div>
                <div className="col-span-7 px-4">
                    <h4 className="text-xl py-4 font-medium">Resultados</h4>
                    {
                        (q === '') ? 
                            <div className="grid grid-cols-1 justify-items-center w-full flex-col bg-slate-100 p-4 rounded">
                                <img src="/assets/searching.svg" alt="" className="w-1/4 mb-3" />
                                <p className="text-center">Realiza una busqueda</p>
                            </div> : (heroes.length === 0) && 
                            <div aria-label="alert-danger"  className="grid grid-cols-1 justify-items-center w-full flex-col bg-red-100 p-4 rounded">
                                <img src="/assets/we-got-a-problem.svg" alt="" className="w-1/4 mb-3" />
                                <p className="text-center">No hay resultados para { q }</p>
                            </div>
                    }
                    {
                        heroes.map(hero => (
                            <HeroCard  key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </>
    )
}
