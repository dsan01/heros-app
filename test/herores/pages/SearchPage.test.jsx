import { SearchPage } from "../../../src/heroes/pages/SearchPage"

const { render, screen, fireEvent } = require("@testing-library/react")
const { MemoryRouter, useSearchParams } = require("react-router-dom")


const mockedSetSearchParams = jest.fn();

jest.mock("react-router-dom", ()=>({
    ...jest.requireActual("react-router-dom"),
    useSearchParams: ()=>[jest.requireActual("react-router-dom").useSearchParams()[0], mockedSetSearchParams ],
}))

describe('Pruebas en <SearchPage>', () => {


    beforeEach(() => jest.clearAllMocks())


    test('Debe de mostrarse correctamente con valores con defecto', () => {
        
        const {container} =render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()
    })

    test('Debe de mostrar a batman y el input con el valor de querystring', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox')
        expect(input.value).toBe('batman')

        const img = screen.getByRole('img')
        expect(img.src).toContain('dc-batman')
    })

    test('Debe de mostar un error si no se encuentra el heroe', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )

        const alert = screen.getByLabelText('alert-danger')
        expect(alert).not.toBeNull()
    })

    test('Debe llamar la funcion con los parametros', () => {
        

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox')
        
        fireEvent.change(input, {target: {name:'searchtext', value: 'superman'}})
        
        const form = screen.getByRole('form')

        fireEvent.submit(form)

        expect( mockedSetSearchParams ).toHaveBeenCalled(); 
        expect( mockedSetSearchParams ).toHaveBeenCalledWith({q:'superman'})
    })
})