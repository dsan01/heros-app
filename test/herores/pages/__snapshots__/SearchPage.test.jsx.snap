// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage> Debe de mostrarse correctamente con valores con defecto 1`] = `
<div>
  <div
    class="grid grid-cols-12 my-5"
  >
    <h1
      class="text-3xl font-semibold col-span-12"
    >
       Buscar un heroe 
    </h1>
    <div
      class="col-span-5 px-4"
    >
      <h4
        class="text-xl py-4 font-medium"
      >
        Busqueda
      </h4>
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class=" shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
          name="searchtext"
          placeholder="Busca un superheroe"
          type="text"
          value=""
        />
        <button
          class="w-max bg-blue-500 px-4 py-2 my-4 rounded-lg text-white cursor-pointer hover:bg-blue-300"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-span-7 px-4"
    >
      <h4
        class="text-xl py-4 font-medium"
      >
        Resultados
      </h4>
      <div
        class="grid grid-cols-1 justify-items-center w-full flex-col bg-slate-100 p-4 rounded"
      >
        <img
          alt=""
          class="w-1/4 mb-3"
          src="/assets/searching.svg"
        />
        <p
          class="text-center"
        >
          Realiza una busqueda
        </p>
      </div>
    </div>
  </div>
</div>
`;
